File: apps/sandbox-runtime/internal/tunnel/agent_channel_test.go
package tunnel

import (
	"testing"

	"github.com/mistlehq/mistle/apps/sandbox-runtime/internal/startup"
)

func TestResolveAgentEndpoint(t *testing.T) {
	t.Run("returns nil endpoint when runtime plan declares zero agent runtimes", func(t *testing.T) {
		endpoint, err := resolveAgentEndpoint([]startup.AgentRuntime{}, []startup.RuntimeClient{})
		if err != nil {
			t.Fatalf("expected endpoint resolution to succeed for zero agent runtimes, got %v", err)
		}
		if endpoint != nil {
			t.Fatal("expected nil endpoint when no agent runtime is declared")
		}
	})

	t.Run("fails when runtime plan declares multiple agent runtimes", func(t *testing.T) {
		_, err := resolveAgentEndpoint([]startup.AgentRuntime{
			{
				BindingID:   "bind_one",
				RuntimeKey:  "runtime_one",
				ClientID:    "client_one",
				EndpointKey: "endpoint_one",
			},
			{
				BindingID:   "bind_two",
				RuntimeKey:  "runtime_two",
				ClientID:    "client_two",
				EndpointKey: "endpoint_two",
			},
		}, []startup.RuntimeClient{
			{
				ClientID: "client_one",
				Endpoints: []startup.RuntimeClientEndpointSpec{
					{
						EndpointKey:    "endpoint_one",
						ConnectionMode: "dedicated",
						Transport: startup.RuntimeClientEndpointTransport{
							Type: "ws",
							URL:  "ws://127.0.0.1:4500",
						},
					},
				},
			},
			{
				ClientID: "client_two",
				Endpoints: []startup.RuntimeClientEndpointSpec{
					{
						EndpointKey:    "endpoint_two",
						ConnectionMode: "dedicated",
						Transport: startup.RuntimeClientEndpointTransport{
							Type: "ws",
							URL:  "ws://127.0.0.1:4600",
						},
					},
				},
			},
		})
		if err == nil {
			t.Fatal("expected endpoint resolution to fail when multiple agent runtimes are declared")
		}
	})

	t.Run("resolves declared agent runtime endpoint", func(t *testing.T) {
		endpoint, err := resolveAgentEndpoint([]startup.AgentRuntime{
			{
				BindingID:   "bind_openai",
				RuntimeKey:  "codex-app-server",
				ClientID:    "client_codex",
				EndpointKey: "app-server",
			},
		}, []startup.RuntimeClient{
			{
				ClientID: "client_codex",
				Endpoints: []startup.RuntimeClientEndpointSpec{
					{
						EndpointKey:    "app-server",
						ConnectionMode: "dedicated",
						Transport: startup.RuntimeClientEndpointTransport{
							Type: "ws",
							URL:  "ws://127.0.0.1:4500",
						},
					},
				},
			},
		})
		if err != nil {
			t.Fatalf("expected endpoint resolution to succeed, got %v", err)
		}
		if endpoint == nil {
			t.Fatal("expected endpoint to be resolved")
		}
		if endpoint.ClientID != "client_codex" {
			t.Fatalf("expected clientId 'client_codex', got '%s'", endpoint.ClientID)
		}
		if endpoint.EndpointKey != "app-server" {
			t.Fatalf("expected endpointKey 'app-server', got '%s'", endpoint.EndpointKey)
		}
		if endpoint.ConnectionMode != "dedicated" {
			t.Fatalf("expected connection mode 'dedicated', got '%s'", endpoint.ConnectionMode)
		}
		if endpoint.TransportURL != "ws://127.0.0.1:4500" {
			t.Fatalf("expected transport url 'ws://127.0.0.1:4500', got '%s'", endpoint.TransportURL)
		}
	})

	t.Run("fails when declared agent endpoint is not websocket", func(t *testing.T) {
		endpoint, err := resolveAgentEndpoint([]startup.AgentRuntime{
			{
				BindingID:   "bind_openai",
				RuntimeKey:  "codex-app-server",
				ClientID:    "client_codex",
				EndpointKey: "http-endpoint",
			},
		}, []startup.RuntimeClient{
			{
				ClientID: "client_codex",
				Endpoints: []startup.RuntimeClientEndpointSpec{
					{
						EndpointKey:    "http-endpoint",
						ConnectionMode: "dedicated",
						Transport: startup.RuntimeClientEndpointTransport{
							Type: "http",
							URL:  "http://127.0.0.1:4500",
						},
					},
				},
			},
		})
		if err == nil {
			t.Fatal("expected endpoint resolution to fail for non-websocket agent endpoint")
		}
		if endpoint != nil {
			t.Fatal("expected no resolved endpoint when agent runtime points at a non-websocket endpoint")
		}
	})
}

File: apps/sandbox-runtime-node/src/tunnel/agent-channel.test.ts
import type { CompiledAgentRuntime, CompiledRuntimeClient } from "@mistle/integrations-core";
import { describe, expect, it } from "vitest";

import { resolveAgentEndpoint } from "./agent-channel.js";

function createRuntimeClient(overrides: Partial<CompiledRuntimeClient>): CompiledRuntimeClient {
  return {
    clientId: "client_openai",
    setup: {
      env: {},
      files: [],
    },
    processes: [],
    endpoints: [
      {
        endpointKey: "app-server",
        transport: {
          type: "ws",
          url: "ws://127.0.0.1:4020/app-server",
        },
        connectionMode: "dedicated",
      },
    ],
    ...overrides,
  };
}

function createAgentRuntime(overrides: Partial<CompiledAgentRuntime>): CompiledAgentRuntime {
  return {
    bindingId: "binding_openai",
    runtimeKey: "codex-app-server",
    clientId: "client_openai",
    endpointKey: "app-server",
    adapterKey: "openai-codex",
    ...overrides,
  };
}

describe("resolveAgentEndpoint", () => {
  it("returns undefined when the runtime plan does not declare an agent runtime", () => {
    expect(resolveAgentEndpoint([], [])).toBeUndefined();
  });

  it("resolves the declared websocket endpoint", () => {
    const resolvedEndpoint = resolveAgentEndpoint(
      [createAgentRuntime({})],
      [createRuntimeClient({})],
    );

    expect(resolvedEndpoint).toEqual({
      runtimeKey: "codex-app-server",
      clientId: "client_openai",
      endpointKey: "app-server",
      connectionMode: "dedicated",
      transportUrl: "ws://127.0.0.1:4020/app-server",
    });
  });

  it("fails when the runtime references a missing client", () => {
    expect(() =>
      resolveAgentEndpoint(
        [createAgentRuntime({ clientId: "missing-client" })],
        [createRuntimeClient({})],
      ),
    ).toThrow("references missing runtime client 'missing-client'");
  });
});
